### PR TITLE
RFC: matches and whenSuccess combinators

### PR DESCRIPTION
**Pull request by mcanlas**
Includes two independent proposals per #614

1) `matches` partial pattern match combinator
2) `existsRight` combinator for asserting `ApplicativeError`

Proposal 2 is more a strawman, I'm not confident in it at all. I had to also demand `Foldable`. I looked at `Bifoldable` briefly but I wasn't sure how to merge the two channels into one. Felt like that was describing two fully independent channels vs a disjoint union in an `ApplicativeError`

### CODE REVIEW

File: modules/core/src/weaver/Expectations.scala
Comment (keynmol) [inline modules/core/src/weaver/Expectations.scala:251]: May be `.attempt.map(_.fold(...))`?

Comment (mcanlas) [reply modules/core/src/weaver/Expectations.scala]: Wasn't totally sure what the goal here would be. Doesn't something need to interpret `F`? In the PR's case, that's what `F : Foldable` is for

`attempt` would still be wrapped in `F`

**Comment by mcanlas** [comment]
Renamed to `whenSuccess`

Wasn't sure about the `Foldable` logic. I realized the left `b` value I was ignoring initially was the accumulator value?

Tried to clarify the logical edge case error messages

What still needs doing (potentially after a thumbs up) is documenting these combinators

**Comment by mcanlas** [comment]
Documentation added!

**Review by Baccata — APPROVED** [review]
Thanks ! I'll wait for @keynmol's feedback before merging.

**Review by keynmol — APPROVED** [review]

**Comment by callado4** [comment]
Any idea when this will be officially released? I would like to use this functionality

**Comment by Baccata** [comment]
Thanks for the reminder. I've triggered a release, it should be here today.

**Pull request merged**
